### ISSUE TITLE
Files_Locking + SQL issue when restoring to an earlier version of a file

### ISSUE THREAD

**Issue by jnfrmarks** [owncloud/core#14591]
### Steps to reproduce

1. Install owncloud 
2. upload or create a text file
3. edit the file a couple of time so that there are a few versions
4. click on the version link for the file
5. select one of the version and click the restore link

### Expected behaviour

The file is restored and the version dropdown goes away

### Actual behaviour

The file is restored but the version dropdown remains.  This makes it seem that the restore didn't happen so I end up clicking it multiple times getting an error message that the file can't be restored because I'm trying to restore a file that has already been restored.

This is a regression from 7.

![p295](https://cloud.githubusercontent.com/assets/8480343/6402096/02fcdb62-bdd3-11e4-9647-3ae0099762b9.PNG)

Server log file contains:

```
Array
(
    [reqId] => 313a0cbcf00ec77ec0cfe5f84736af63
    [remoteAddr] => x.x.x.x
    [app] => index
    [message] => Exception:
    	Array
		(
		    [Message] => An exception occurred while executing 'UPDATE `filecache` SET `path` = ?, `path_hash` = ?, `name` = ?, `parent` =? WHERE `fileid` = ?' with params ["files\/pre_141.txt", "a296a454c71cf3abb1d4c8d60dfb151d", "pre_141.txt", "3", 40]:

		SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '1-a296a454c71cf3abb1d4c8d60dfb151d' for key 'fs_storage_path_hash'
		    [Code] => 0
		    [Trace] => #0 /var/www/html/owncloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php(116): Doctrine\DBAL\Driver\AbstractMySQLDriver->convertException('An exception oc...', Object(Doctrine\DBAL\Driver\PDOException))
		#1 /var/www/html/owncloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Statement.php(172): Doctrine\DBAL\DBALException::driverExceptionDuringQuery(Object(Doctrine\DBAL\Driver\PDOMySql\Driver), Object(Doctrine\DBAL\Driver\PDOException), 'UPDATE `filecac...', Array)
		#2 /var/www/html/owncloud/lib/private/db/statementwrapper.php(63): Doctrine\DBAL\Statement->execute(Array)
		#3 /var/www/html/owncloud/lib/private/db.php(141): OC_DB_StatementWrapper->execute(Array)
		#4 /var/www/html/owncloud/lib/private/files/cache/cache.php(424): OC_DB::executeAudited('UPDATE `*PREFIX...', Array)
		#5 /var/www/html/owncloud/lib/private/files/cache/updater.php(107): OC\Files\Cache\Cache->move('files_versions/...', 'files/pre_141.t...')
		#6 /var/www/html/owncloud/lib/private/files/view.php(546): OC\Files\Cache\Updater->rename('/files_versions...', '/files/pre_141....')
		#7 /var/www/html/owncloud/apps/files_versions/lib/storage.php(273): OC\Files\View->rename('files_versions/...', 'files/pre_141.t...')
		#8 /var/www/html/owncloud/apps/files_versions/ajax/rollbackVersion.php(10): OCA\Files_Versions\Storage::rollback('/pre_141.txt', 1425059785)
		#9 /var/www/html/owncloud/lib/private/route/route.php(135) : runtime-created function(1): require_once('/var/www/html/o...')
		#10 [internal function]: __lambda_func(Array)
		#11 /var/www/html/owncloud/lib/private/route/router.php(250): call_user_func('?lambda_1158', Array)
		#12 /var/www/html/owncloud/lib/base.php(767): OC\Route\Router->match('/apps/files_ver...')
		#13 /var/www/html/owncloud/index.php(36): OC::handleRequest()
		#14 {main}
		    [File] => /var/www/html/owncloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php
		    [Line] => 66
		)
    [level] => 4
    [time] => 2015-02-27T17:57:15+00:00
)

```

### Server Host info:

CentOS release 6.6 (Final)
PHP 5.4.36

### Server configuration

ownCloud Enterprise Edition 8.0 (daily) Build:2015-02-27T04:33:31+00:00 5faf9f819212eb0d2321fe324831b941a215819e plus patch for 1458

**Comment by jnfrmarks** [comment]
@LukasReschke @PVince81 

As requested ...

**Comment by PVince81** [comment]
@icewind1991 any idea ?

**Comment by jnfrmarks** [comment]
More from the log file:

```

{"reqId":"97a560af79b62fa3b57bd67502924116","remoteAddr":"192.168.1.218","app":"index","message":"Exception: {\"Message\":\"Could not obtain lock type 1 on \\\"\\\/var\\\/www\\\/html\\\/owncloud\\\/data\\\/at_admin\\\/files_versions\\\/pre_138.txt.v1425055498\\\".\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/html\\\/owncloud\\\/apps\\\/files_locking\\\/lib\\\/lockingwrapper.php(44): OCA\\\\Files_Locking\\\\Lock->addLock(1, NULL)\\n#1 \\\/var\\\/www\\\/html\\\/owncloud\\\/apps\\\/files_locking\\\/lib\\\/lockingwrapper.php(126): OCA\\\\Files_Locking\\\\LockingWrapper->getLock('files_versions\\\/...', 1)\\n#2 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/wrapper.php(244): OCA\\\\Files_Locking\\\\LockingWrapper->rename('files_versions\\\/...', 'files\\\/pre_138.t...')\\n#3 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(512): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->rename('files_versions\\\/...', 'files\\\/pre_138.t...')\\n#4 \\\/var\\\/www\\\/html\\\/owncloud\\\/apps\\\/files_versions\\\/lib\\\/storage.php(273): OC\\\\Files\\\\View->rename('files_versions\\\/...', 'files\\\/pre_138.t...')\\n#5 \\\/var\\\/www\\\/html\\\/owncloud\\\/apps\\\/files_versions\\\/ajax\\\/rollbackVersion.php(10): OCA\\\\Files_Versions\\\\Storage::rollback('\\\/pre_138.txt', 1425055498)\\n#6 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/route\\\/route.php(135) : runtime-created function(1): require_once('\\\/var\\\/www\\\/html\\\/o...')\\n#7 [internal function]: __lambda_func(Array)\\n#8 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/route\\\/router.php(250): call_user_func('?lambda_1158', Array)\\n#9 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/base.php(767): OC\\\\Route\\\\Router->match('\\\/apps\\\/files_ver...')\\n#10 \\\/var\\\/www\\\/html\\\/owncloud\\\/index.php(36): OC::handleRequest()\\n#11 {main}\",\"File\":\"\\\/var\\\/www\\\/html\\\/owncloud\\\/apps\\\/files_locking\\\/lib\\\/lock.php\",\"Line\":246}","level":4,"time":"2015-02-27T17:55:51+00:00"}

```


Note that this is a different file than the one above

**Comment by jnfrmarks** [comment]
App list

```
ubuntu-14-lts:/var/www/html/owncloud# sudo -su www-data ./occ app:list
Enabled:
 - activity
 - admin_audit
 - enterprise_key
 - files
 - files_locking
 - files_pdfviewer
 - files_sharing
 - files_sharing_log
 - files_texteditor
 - files_trashbin
 - files_versions
 - files_videoviewer
 - firewall
 - gallery
 - provisioning_api
 - templateeditor
 - updater
 - windows_network_drive
Disabled:
 - external
 - files_antivirus
 - files_drop
 - files_encryption
 - files_external
 - files_ldap_home
 - firstrunwizard
 - sharepoint
 - user_external
 - user_ldap
 - user_shibboleth
 - user_webdavauth
```

**Comment by LukasReschke** [comment]
I beautified the error messages a little bit.

**Comment by LukasReschke** [comment]
@icewind1991 Please take a look – THX…

**Comment by LukasReschke** [comment]
(can't reproduce the SQL problem on stable8. The files_locking application seems like some kind of race-condition?)

**Comment by PVince81** [comment]
Works for me with enterprise 8.0.1 RC1 and MariaDB 10.
Restoring the version works (the thumbnail is not updated, but that's another issue)

@jnfrmarks what database did you use ? According to the error message it's either MySQL or MariaDB.

**Comment by LukasReschke** [comment]
I guess the SQL error is a problem caused by a wrong applied patch (I guess I gave @jnfrmarks the wrong commands … my bad)

What is left is that files_locking problem then :imp:

**Comment by LukasReschke** [comment]
(That said I also can't reproduce that – I suspect a race condition)

**Comment by jnfrmarks** [comment]
Resolved in 8.0.x
